### PR TITLE
Fixed issue #20271: Reflected XSS in Lime Survey Version 6.15.10+250901

### PR DESCRIPTION
**Pull request by gabrieljenik**
### **PR Type**
Bug fix

___

### **Description**
- Fixed reflected XSS vulnerability in CSRF token handling

- Added CSRF token sanitization to prevent malicious script injection

- Implemented comprehensive unit tests for CSRF validation scenarios

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["CSRF Cookie"] --> B["getCsrfToken()"]
  B --> C["sanitize_csrf_token()"]
  C --> D["Sanitized Token"]
  D --> E["XSS Prevention"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>LSHttpRequest.php</strong><dd><code>Override CSRF token method with sanitization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

application/core/LSHttpRequest.php

<ul><li>Added protected <code>_csrfToken</code> property for token caching<br> <li> Overrode <code>getCsrfToken()</code> method to sanitize CSRF tokens<br> <li> Integrated <code>sanitize_csrf_token()</code> function to prevent XSS attacks</ul>

</details>

  </td>
  <td><a href="https://github.com/LimeSurvey/LimeSurvey/pull/4452/files#diff-51e6b3df6b903a53bafb6ef5152bd9d3b6d150502cb833021ef38a3a349fb795">+20/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>sanitize_helper.php</strong><dd><code>Add CSRF token sanitization helper function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

application/helpers/sanitize_helper.php

<ul><li>Added <code>sanitize_csrf_token()</code> function<br> <li> Removes invalid characters from CSRF tokens using regex<br> <li> Allows only alphanumeric, underscore, hyphen, and equals characters</ul>

</details>

  </td>
  <td><a href="https://github.com/LimeSurvey/LimeSurvey/pull/4452/files#diff-ac66a40bc5e4438b4e825f82a53dbcb0d0d2bd5e617654ee904c1889e9fc1b8e">+12/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CsrfHttpRequestTest.php</strong><dd><code>Add comprehensive CSRF token validation tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/unit/CsrfHttpRequestTest.php

<ul><li>Added <code>testCsrfTokenValidation()</code> method with data provider<br> <li> Tests valid and invalid CSRF tokens with sanitization scenarios<br> <li> Covers edge cases including XSS payloads and special characters<br> <li> Validates both token sanitization and CSRF validation behavior</ul>

</details>

  </td>
  <td><a href="https://github.com/LimeSurvey/LimeSurvey/pull/4452/files#diff-1c68b571850458ceabdad5e59da33779f5eefb95ae32a7f616101cd92aa391c0">+113/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

**Comment by gabrieljenik** [comment]
The issue was on the sanitization. It was not a pure base64 string but it was masked after. So some characters which may be eventually part of the token could have been stripped off.
Added new unit tests for the whole sanitization as well as for the validation.

**Comment by gabrieljenik** [comment]
Something that crosses my mind:
- If the cookie value has invalid characters, should we just clear the cookie so all validations fail?
- Or the other way around, if a cookie is actually sanitized, should we update the cookie value in case some else is picking it up?

File: application/core/LSHttpRequest.php [hunk 1]
```diff
@@ -313,4 +315,22 @@ public static function checkIsAllowedHost($hostInfo)
...
+
+    /**
+     * @inheritdoc
+     * Sanitize the CSRF token to avoid XSS attacks
+     */
+    public function getCsrfToken()
+    {
+        if ($this->_csrfToken === null) {
+            $cookie = $this->getCookies()->itemAt($this->csrfTokenName);
+            if (!$cookie || ($this->_csrfToken = sanitize_csrf_token($cookie->value)) == null) {
```
Comment (Shnoulle) [inline application/core/LSHttpRequest.php]: Why not 

```
if ($cookie && $cookie->value != sanitize_csrf_token($cookie->value)) {
throw new CHttpException(400);
}
```

Then : sysadmin can track it a little (get date time etc …)

**Comment by Shnoulle** [comment]
> If the cookie value has invalid characters, should we just clear the cookie so all validations fail?

Clearly ! Attack done somewhere ! We can not use such cookie value !

File: application/core/LSHttpRequest.php (hunk 1 above)
Comment (Shnoulle) [reply application/core/LSHttpRequest.php]: With clearing cookie 

```
if ($cookie && $cookie->value != sanitize_csrf_token($cookie->value)) {
    unset(Yii::app()->request->cookies[$this->csrfTokenName]);
    throw new CHttpException(400);
}
```

**Review by c-schmitz — CHANGES_REQUESTED** [review]
The patch does not fix at all the unsafe output of the CSRFToken at all.
I think we need to go back here and fix that one first - one location as pointed out by 
Denis already is /application/extensions/LimeScript/LimeScript.php.

I am not sure at what location the original bug displays the  malicious CSRF token, but fixing that improper output should be top priority:

A possible fix for LimeScript could be (untested):

```
// ... existing code
            $json = CJavaScript::encode($data);
            $script = "LS.data = $json;\n
                    // @see https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#jquery
                    function csrfSafeMethod(method) {
                        // these HTTP methods do not require CSRF protection
                        return (/^(GET|HEAD|OPTIONS|TRACE)$/.test(method));
                    }
                    $.ajaxSetup({
                        beforeSend: function(xhr, settings) {
                            if (!csrfSafeMethod(settings.type) && !this.crossDomain) {
                                if (typeof settings.data == 'undefined') {
                                    settings.data = '';
                                }
                                // Use the safe LS.data object instead of PHP concatenation
                                if (settings.data.indexOf(LS.data.csrfTokenName) == -1)
                                {
                                    if (settings.data != '') {
                                        settings.data += '&';
                                    }
                                    settings.data += LS.data.csrfTokenName + '=' + LS.data.csrfToken;
                                }
                            }
                        }
                    });
            ";
            App()->getClientScript()->registerScript('LimeScript', $script, CClientScript::POS_HEAD);
        }
    }
```

**Comment by gabrieljenik** [comment]
@Shnoulle We have just updated the PR as to throw an exception as soon as an invalid token is found.

@c-schmitz  regaridn the following you mentioned....

> The patch does not fix at all the unsafe output of the CSRFToken at all. I think we need to go back here and fix that one first - one location as pointed out by Denis already is /application/extensions/LimeScript/LimeScript.php.

I guess you are talking about a comment Denis left on the ticket (https://bugs.limesurvey.org/view.php?id=20271#c83468)

I am not sure if that comment still holds.
As of the approach on this PR, as soon as we identify an invalid CSRF token, an exception is being thrown.
So from my perspective, that part of the code will not get to be run.

What do you think?

Thanks!

**Comment by Shnoulle** [comment]
> I guess you are talking about a comment Denis left on the ticket (https://bugs.limesurvey.org/view.php?id=20271#c83468)
> 
> I am not sure if that comment still holds. As of the approach on this PR, as soon as we identify an invalid CSRF token, an exception is being thrown. So from my perspective, that part of the code will not get to be run.


My opinion here : Yii dev choose to allow any CRSF but encode since it send by user. It's the solution "_Do not trust what user sent, but do not update it_". LimeSurvey didn't encode it when using it for script, encode it use the same solution.

I like the "_Do not trust what user sent, but do not update it_" solution.

**Comment by Shnoulle** [comment]
> A possible fix for LimeScript could be (untested):

My opinion : 
Check if it works for all request , replace https://github.com/LimeSurvey/LimeSurvey/blob/fd91b5747369b4b52849c9257c73d1d62a42c6b1/application/extensions/LimeScript/LimeScript.php#L17

```
            $data['csrfTokenName']              = Chtml::encode(Yii::app()->request->csrfTokenName);
            $data['csrfToken']                  = Chtml::encode(Yii::app()->request->csrfToken);
```

Since it uses ENT_QUOTES : replace `'` and `"`

**Comment by gabrieljenik** [comment]
> I like the "_Do not trust what user sent, but do not update it_" solution.

We are not updating it. With the new commit from yesterday, we are throwing an exception.
So, if I understand correctly, no further fixes are needed.

> A possible fix for LimeScript could be (untested):

That is not needed as the page flow will never get to that. Are you able to reproduce the issue?
Also, encoding the token implies checking not only there but I think in other different places where the token is used.

**Comment by Shnoulle** [comment]
> That is not needed as the page flow will never get to that. Are you able to reproduce the issue? Also, encoding the token implies checking not only there but I think in other different places where the token is used.

Maybe, but it's already encoded by Yii :)

Some grep to find where ?

**Comment by gabrieljenik** [comment]
> Maybe, but it's already encoded by Yii :)
> 
> Some grep to find where ?

Why spending time on that and widening the possibility of new errors?

**Comment by Shnoulle** [comment]
> > Maybe, but it's already encoded by Yii :)
> > Some grep to find where ?
> 
> Why spending time on that and widening the possibility of new errors?

@c-schmitz  decision here :)

**Comment by c-schmitz** [comment]
The extensive validation of the CSRF key seems excessive and inefficient. 
It's sufficient to ensure the key is properly quoted when rendered in HTML to prevent breaking the markup and triggering XSS vulnerabilities.
Please update the patch accordingly.

**Comment by gabrieljenik** [comment]
Closed in favor of https://github.com/LimeSurvey/LimeSurvey/pull/4518
